crates/mempool/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright 2024 PRAGMA
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+use pallas_primitives::conway::{TransactionInput, Tx};
+use amaru_ledger::store::{Snapshot};
+
+pub trait Mempool {
+    type Tx;
+    type Block;
+    fn add_tx<S: Snapshot>(&mut self, snapshot: &S, tx: Self::Tx) -> bool;
+    fn make_block<S: Snapshot>(&mut self, snapshot: &S) -> Option<Self::Block>;
+    fn invalidate_utxos(&mut self, txins: HashSet<TransactionInput>);
```

Comment:
similarly, rather than thinking of this as `invalidate_utxo`, we probably want to provide a new snapshot, or some delta to the last snapshot. For example, if a block gets minted that deregisters a stake address, that might invalidate a transaction that is trying to withdraw the staking rewards.

For performance reasons, we likely don't want to re-validate all of the existing transactions from scratch using the whole ledger snapshot, so that's why I said a delta, but conceptually that's what we're doing.